dlib/test/dnn.cpp

```diff
@@ -2555,6 +2555,202 @@ namespace
+
+    void test_loss_binary_log_per_pixel_learned_params_on_trivial_two_pixel_task()
+    {
+        print_spinner();
+
+        ::std::vector<matrix<float>> x({ matrix<float,2,1>({ -1, 1 }) });
+        ::std::vector<matrix<float>> y({ matrix<float,2,1>({ -1, 1 }) });
+
+        using net_type = loss_binary_log_per_pixel<con<1,1,1,1,1,input<matrix<float>>>>;
+        net_type net;
+
+        dnn_trainer<net_type> trainer(net, sgd(0,0));
+        trainer.set_learning_rate(1e7);
+        trainer.set_max_num_epochs(1);
+        trainer.train(x, y);
+
+        const tensor& learned_params = layer<1>(net).layer_details().get_layer_params();
+        const float* learned_params_data = learned_params.host();
+
+        DLIB_TEST(learned_params_data[0] > 1e5);
+        DLIB_TEST(abs(learned_params_data[1]) < 1);
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    void test_loss_binary_log_per_pixel_outputs_on_trivial_task()
+    {
+        print_spinner();
+
+        constexpr int input_height = 7;
+        constexpr int input_width = 5;
+        constexpr int output_height = input_height;
+        constexpr int output_width = input_width;
+        constexpr int num_samples = 7;
+        constexpr int filter_height = 3;
+        constexpr int filter_width = 3;
+
+        ::std::vector<matrix<double>> x(num_samples);
+        ::std::vector<matrix<float>> y(num_samples);
+
+        matrix<double> xtmp(input_height, input_width);
+        matrix<float> ytmp(output_height, output_width);
+
+        ::std::default_random_engine generator(16);
+        ::std::normal_distribution<double> n01(0);
+
+        const auto z = 0.674490; // This should give us a 50/50 split between the classes
+
+        // Generate training data: random inputs x, and the corresponding target outputs y
+        for (int ii = 0; ii < num_samples; ++ii) {
+            for (int jj = 0; jj < input_height; ++jj) {
+                for (int kk = 0; kk < input_width; ++kk) {
+                    xtmp(jj, kk) = n01(generator);
+                    ytmp(jj, kk) = std::abs(xtmp(jj, kk)) > z ? 1.f : -1.f;
+                }
+            }
+            x[ii] = xtmp;
+            y[ii] = ytmp;
+        }
+
+        using net_type = loss_binary_log_per_pixel<con<1,1,1,1,1,relu<con<10,1,1,1,1,input<matrix<double>>>>>>;
+        net_type net;
+
+        dnn_trainer<net_type> trainer(net, sgd(0, 0.9));
+        trainer.set_learning_rate(1);
+        trainer.set_max_num_epochs(200);
+        trainer.train(x, y);
+
+        // The learning task is easy, so the net should have no problem
+        // getting all the outputs right.
+        const auto response = net(x);
+        for (int ii = 0; ii < num_samples; ++ii)
+            for (int jj = 0; jj < output_height; ++jj)
+                for (int kk = 0; kk < output_width; ++kk)
+                    DLIB_TEST((response[ii](jj,kk) > 0) == (y[ii](jj,kk) > 0));
```

Comment:
I'm getting a failure here when I run this test with dlib built to use cuda.  Is it working for you?

Reply:
Yes, it works on my machine (CUDA build).

Tried dozens of times now, but couldn't get a failure.

Doubled the number of epochs to train. Does that help?

Reply:
I am also getting an error with 400 epochs. Doubling it works for me, but also setting it 500.

Reply:
Thanks @arrufat. On my machine it worked with 200, and the 400 was already a doubled-up number. Now increased further to 800.

Reply:
Cool, working for me now too :)